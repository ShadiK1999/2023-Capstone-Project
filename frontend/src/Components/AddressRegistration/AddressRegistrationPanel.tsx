import * as styles from './AddressRegistration.styles';
import { FC, useCallback, useState, useRef, useEffect } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { Autocomplete } from '@react-google-maps/api';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';
import { DEFAULT_BOUNDS } from '../../Services/AppConstants';
import { DayOfWeekButton, TimeOfDayButton, LabelBox } from '../';
import { DayOfWeekIndex } from '../../Contexts/Home/state';
import { AddressPost } from '../../Services/User';
import { OpeningHours, TrustedPoint } from '../../Models/DeliveryPoint';

export const errorMessages = {
  address: 'Address is required.',
  daysOfWeek: 'Please select at least one day.',
};

const { address, daysOfWeek } = errorMessages;

const validationSchema = z.object({
  address: z.string().nonempty({ message: address }),
  daysOfWeek: z.array(z.number()).min(1, { message: daysOfWeek }),
});

type ValidationSchema = z.infer<typeof validationSchema>;
const resolver = zodResolver(validationSchema);

type ComponentProps = {
  initialAddress?: TrustedPoint;
  onClose: () => void;
  onEnter: (_: AddressPost) => Promise<void>;
};

const AddressRegistrationPanel: FC<ComponentProps> = ({ onClose, onEnter, initialAddress }) => {
  const searchRef = useRef<google.maps.places.Autocomplete | null>(null);
  const [latlng, setLatlng] = useState<google.maps.LatLngLiteral | null>(
    initialAddress?.location ?? null,
  );
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<ValidationSchema>({ resolver });
  const [submitError, setSubmitError] = useState<string | null>(null);
  const [daysOfWeek, setDaysOfWeek] = useState<DayOfWeekIndex[]>(
    Object.keys(initialAddress?.hours ?? {}).map((hour) => +hour as DayOfWeekIndex),
  );
  const [startTime, setStartTime] = useState<{ hour: number; minute: number }>({
    hour: 9,
    minute: 0,
  });
  const [endTime, setEndTime] = useState<{ hour: number; minute: number }>({
    hour: 17,
    minute: 0,
  });

  useEffect(() => {
    setValue('daysOfWeek', daysOfWeek);
  }, [setValue, daysOfWeek]);

  const onSubmit: SubmitHandler<ValidationSchema> = useCallback(
    async (data) => {
      if (latlng) {
        const hours: OpeningHours = {};
        daysOfWeek.forEach((day) => {
          hours[day] = {
            openingTime: startTime,
            closingTime: endTime,
          };
        });
        const body: AddressPost = {
          address: data.address,
          location: latlng,
          hours: hours,
        };

        await onEnter(body);
        onClose();
      } else {
        setSubmitError('Please check your address again');
      }
    },
    [latlng, daysOfWeek, onEnter, onClose, startTime, endTime],
  );

  const onLoad = useCallback((searchBox: google.maps.places.Autocomplete) => {
    searchRef.current = searchBox;
  }, []);

  const onSearch = useCallback(() => {
    if (!searchRef.current) return;
    const place = searchRef.current.getPlace();
    const latlng = place?.geometry?.location;
    if (!latlng) return;
    setLatlng({
      lat: latlng.lat(),
      lng: latlng.lng(),
    });
  }, []);

  return (
    <div className={styles.Component}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className={styles.InputContainer}>
          <LabelBox label="Address">
            <Autocomplete onLoad={onLoad} onPlaceChanged={onSearch} bounds={DEFAULT_BOUNDS}>
              <input
                type="text"
                data-testid="address"
                {...register('address')}
                value={initialAddress?.address}
                readOnly={!!initialAddress}
              />
            </Autocomplete>
          </LabelBox>
          {errors?.address && <span>{errors.address.message}</span>}
        </div>
        <div className={styles.InputContainer}>
          <DayOfWeekButton
            daysOfWeek={daysOfWeek}
            onUpdateDaysOfWeek={setDaysOfWeek}
            multiSelect={true}
          />
          {errors?.daysOfWeek && <span>{errors.daysOfWeek.message}</span>}
        </div>
        <TimeOfDayButton
          startTime={startTime}
          endTime={endTime}
          onUpdateStartTime={setStartTime}
          onUpdateEndTime={setEndTime}
        />
        <p>
          For address verification, our team will contact you within 48 hours to confirm your
          residency.
        </p>
        <div className={styles.InputContainer}>
          <button type="submit">Submit</button>
          {submitError && <span>{submitError}</span>}
        </div>
      </form>
    </div>
  );
};

export { AddressRegistrationPanel };
