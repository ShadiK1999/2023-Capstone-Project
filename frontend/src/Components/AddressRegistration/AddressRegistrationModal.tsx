import { Modal } from '..';
import { FC } from 'react';
import { AddressRegistrationPanel } from './AddressRegistrationPanel';
import { TrustedPoint } from '../../Models/DeliveryPoint';
import { AddressPost } from '../../Services/User';

type ComponentProps = {
  isOpen: boolean;
  initialAddress?: TrustedPoint;
  onClose: () => void;
  onEnter: (_: AddressPost) => Promise<void>;
};

const AddressRegistrationModal: FC<ComponentProps> = ({
  isOpen,
  onClose,
  onEnter,
  initialAddress,
}) => {
  return (
    <Modal title="Add Address Details" isOpen={isOpen} onClose={onClose}>
      <AddressRegistrationPanel
        onClose={onClose}
        onEnter={onEnter}
        initialAddress={initialAddress}
      />
    </Modal>
  );
};

export { AddressRegistrationModal };
