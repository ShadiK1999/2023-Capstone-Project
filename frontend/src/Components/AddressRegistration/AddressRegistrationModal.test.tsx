import { render, fireEvent } from '@testing-library/react';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { AddressRegistrationModal } from './AddressRegistrationModal';
import { errorMessages } from './AddressRegistrationPanel';

vi.mock('../../App', () => ({
  router: () => {},
}));

describe('Matches DOM Snapshot', () => {
  const { address, daysOfWeek } = errorMessages;

  beforeEach(() => {
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    if (modalRoot) {
      document.body.removeChild(modalRoot);
    }
  });

  it('Matches DOM Snapshot', () => {
    const tree = render(
      <AddressRegistrationModal
        isOpen={true}
        onClose={() => undefined}
        onEnter={() => Promise.resolve()}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('Does not allow empty field values', async () => {
    const { getByText, findByText } = render(
      <AddressRegistrationModal
        isOpen={true}
        onClose={() => undefined}
        onEnter={() => Promise.resolve()}
      />,
    );
    fireEvent.click(getByText('Submit'));
    expect(await findByText(address)).toBeInTheDocument();
    expect(await findByText(daysOfWeek)).toBeInTheDocument();
  });
});
