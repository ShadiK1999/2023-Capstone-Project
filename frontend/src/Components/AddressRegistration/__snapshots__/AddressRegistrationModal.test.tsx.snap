// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches DOM Snapshot > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="modal-root"
    >
      <div
        class="f1atpta9"
      >
        <h1>
          Add Address Details
        </h1>
        <div
          class="f2vhqpx"
        >
          <form>
            <div
              class="f1k1m9me"
            >
              <div
                class="f1vzetsa"
              >
                <p>
                  Address
                </p>
                <div
                  class="f1nym8no"
                >
                  <div
                    class=""
                  >
                    <input
                      data-testid="address"
                      name="address"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="f1k1m9me"
            >
              <div
                class="f1vzetsa"
              >
                <p>
                  Open on
                </p>
                <div
                  class="f1nym8no"
                >
                  <div
                    class="f1o2virz"
                  >
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      sun
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      mon
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      tue
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      wed
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      thu
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      fri
                    </div>
                    <div
                      class="f10rx374"
                      data-selected="false"
                    >
                      sat
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="f1vzetsa"
            >
              <p>
                Open from
              </p>
              <div
                class="f1nym8no"
              >
                <div
                  class="f723edx"
                >
                  <div
                    class="f2ey5g1"
                  >
                    <img
                      alt="MINUS"
                      src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                    />
                  </div>
                  <div
                    class="f1sz7xol"
                  >
                    09
                    :
                    00
                  </div>
                  <div
                    class="f1fozamg"
                  >
                    <img
                      alt="PLUS"
                      src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="f1vzetsa"
            >
              <p>
                Open until
              </p>
              <div
                class="f1nym8no"
              >
                <div
                  class="f723edx"
                >
                  <div
                    class="f2ey5g1"
                  >
                    <img
                      alt="MINUS"
                      src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                    />
                  </div>
                  <div
                    class="f1sz7xol"
                  >
                    17
                    :
                    00
                  </div>
                  <div
                    class="f1fozamg"
                  >
                    <img
                      alt="PLUS"
                      src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                    />
                  </div>
                </div>
              </div>
            </div>
            <p>
              For address verification, our team will contact you within 48 hours to confirm your residency.
            </p>
            <div
              class="f1k1m9me"
            >
              <button
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
        <a
          href="#"
        >
          Close
        </a>
      </div>
    </div>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
